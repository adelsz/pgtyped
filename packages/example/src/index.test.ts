import { afterEach, beforeAll, beforeEach, expect, test } from '@jest/globals';
import pg from 'pg';
import {
  aggregateEmailsAndTest,
  countBooks,
  findBookById,
  findBookNameOrRank,
  findBookUnicode,
  getBooks,
  getBooksByAuthorName,
  insertBooks,
  updateBooks,
  updateBooksCustom,
  updateBooksRankNotNull,
} from './books/books.queries.js';
import {
  getAllComments,
  insertComment,
  selectExistsTest,
} from './comments/comments.queries.js';
import {
  insertNotification,
  insertNotifications,
} from './notifications/notifications.js';
import {
  getNotifications,
  sendNotifications,
  thresholdFrogs,
} from './notifications/notifications.queries.js';
import { getUsersWithComment } from './users/sample.js';
import { Category } from './customTypes';

const { Client } = pg;

const dbConfig = {
  host: process.env.PGHOST ?? '127.0.0.1',
  user: process.env.PGUSER ?? 'postgres',
  password: process.env.PGPASSWORD ?? 'password',
  database: process.env.PGDATABASE ?? 'postgres',
  port: (process.env.PGPORT ? Number(process.env.PGPORT) : undefined) ?? 5432,
};

// Connect to the database once before all tests
let client: pg.Client;
beforeAll(async () => {
  // Parse dates as strings for demo and testing purposes
  pg.types.setTypeParser(pg.types.builtins.DATE, function (val) {
    return val;
  });

  pg.types.setTypeParser(pg.types.builtins.INT8, function (val) {
    return BigInt(val);
  });

  // Create a new client and connect to the database
  client = new Client(dbConfig);
  await client.connect();
});

// Disconnect from the database after all tests
afterAll(async () => {
  await client.end();
});

// Run each test in a transaction that is rolled back at the end
beforeEach(() => client.query('BEGIN'));
afterEach(() => client.query('ROLLBACK'));

test('select query with unicode characters', () => {
  const result = findBookUnicode.run(undefined, client);
  expect(result).resolves.toMatchSnapshot();
});

test('select query with parameters', async () => {
  const comments = await getAllComments.run({ id: 1 }, client);
  expect(comments).toMatchSnapshot();
});

test('select query with dynamic or', () => {
  const result = findBookNameOrRank.run(
    {
      rank: 1,
    },
    client,
  );
  expect(result).resolves.toMatchSnapshot();
});

test('select query with date type override (TS)', async () => {
  const comments = await getUsersWithComment(0, client);
  const dateAsString: string = comments.registration_date;
  expect(typeof dateAsString).toBe('string');
});

test('select query with date type override (SQL)', async () => {
  const notifications = await getNotifications.run(
    { userId: 1, date: '2000-01-01' },
    client,
  );
  const dateAsString: string = notifications[0].created_at;
  expect(typeof dateAsString).toBe('string');
});

test('insert query with parameter spread', async () => {
  const [{ book_id: insertedBookId }] = await insertBooks.run(
    {
      books: [
        {
          authorId: 1,
          name: 'A Brief History of Time: From the Big Bang to Black Holes',
          rank: 1,
          categories: [Category.Novel, Category.ScienceFiction],
        },
      ],
    },
    client,
  );
  const { 0: insertedBook } = await findBookById.run(
    { id: insertedBookId },
    client,
  );
  expect(insertedBook.categories).toEqual('{novel,science-fiction}');
});

test('update query with a non-null parameter override', async () => {
  await updateBooks.run({ id: 2, rank: 12, name: 'Another title' }, client);
});

test('insert query with an inline sql comment', async () => {
  const [result] = await insertComment.run(
    { comments: [{ commentBody: 'Just a comment', userId: 1 }] },
    client,
  );
  expect(result).toMatchSnapshot({
    id: expect.any(Number),
  });
});

test('dynamic update query', async () => {
  await updateBooksCustom.run({ id: 2, rank: 13 }, client);
});

test('update query with a multiple non-null parameter overrides', async () => {
  await updateBooksRankNotNull.run(
    { id: 2, rank: 12, name: 'Another title' },
    client,
  );
});

test('select query with join and a parameter override', async () => {
  const books = await getBooksByAuthorName.run(
    {
      authorName: 'Carl Sagan',
    },
    client,
  );
  expect(books).toMatchSnapshot();
});

test('select query with aggregation', async () => {
  const [aggregateData] = await aggregateEmailsAndTest.run(
    { testAges: [35, 23, 19] },
    client,
  );
  expect(aggregateData.agetest).toBe(true);
  expect(aggregateData.emails).toEqual([
    'alex.doe@example.com',
    'jane.holmes@example.com',
    'andrewjackson@example.com',
  ]);
});

test('insert query with an enum field', async () => {
  await sendNotifications.run(
    {
      notifications: [
        {
          user_id: 2,
          payload: { num_frogs: 82 },
          type: 'reminder',
        },
      ],
    },
    client,
  );
});

test('multiple insert queries with an enum field', async () => {
  await insertNotifications.run(
    {
      params: [
        {
          user_id: 1,
          payload: { num_frogs: 1002 },
          type: 'reminder',
        },
      ],
    },
    client,
  );
  await insertNotification.run(
    {
      notification: {
        user_id: 1,
        payload: { num_frogs: 1002 },
        type: 'reminder',
      },
    },
    client,
  );
});

test('select query with json fields and casts', async () => {
  const notifications = await thresholdFrogs.run({ numFrogs: 80 }, client);
  expect(notifications).toMatchSnapshot();
});

test('select query nullability override on return field', async () => {
  const result = await getBooks.run(undefined, client);
  expect(result).toMatchSnapshot();
});

test('select exists query, testing #472', async () => {
  const result = await selectExistsTest.run(undefined, client);
  expect(result).toMatchSnapshot();
});

test('select query with a bigint field', async () => {
  const [row] = await countBooks.run(undefined, client);
  expect(typeof row.book_count).toBe('bigint');
  expect(row.book_count).toBe(BigInt(4));
});
