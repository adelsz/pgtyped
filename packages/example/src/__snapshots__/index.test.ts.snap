// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert query with an inline sql comment 1`] = `
Object {
  "body": "Just a comment",
  "book_id": null,
  "id": Any<Number>,
  "user_id": 1,
}
`;

exports[`select exists query, testing #472 1`] = `
Array [
  Object {
    "isTransactionExists": true,
  },
]
`;

exports[`select query nullability override on return field 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Black Swan",
  },
  Object {
    "id": 2,
    "name": "The Dragons Of Eden",
  },
  Object {
    "id": 3,
    "name": "Mysteries of a Barbershop",
  },
  Object {
    "id": 4,
    "name": "In the Jungle of Cities",
  },
]
`;

exports[`select query with dynamic or 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Black Swan",
  },
]
`;

exports[`select query with interval field 1`] = `
Array [
  Object {
    "duration": "0 years 0 mons 0 days 1 hours 23 mins 45.0 secs",
    "id": 1,
    "text_book_id": 1,
  },
  Object {
    "duration": "0 years 0 mons 0 days 2 hours 34 mins 56.0 secs",
    "id": 2,
    "text_book_id": 2,
  },
  Object {
    "duration": "0 years 0 mons 0 days 3 hours 45 mins 57.0 secs",
    "id": 3,
    "text_book_id": 3,
  },
  Object {
    "duration": "0 years 0 mons 0 days 4 hours 56 mins 48.0 secs",
    "id": 4,
    "text_book_id": 4,
  },
]
`;

exports[`select query with join and a parameter override 1`] = `
Array [
  Object {
    "author_id": 2,
    "categories": null,
    "id": 2,
    "name": "The Dragons Of Eden",
    "rank": 4,
  },
]
`;

exports[`select query with json fields and casts 1`] = `Array []`;

exports[`select query with parameters 1`] = `
Array [
  Object {
    "body": "Fantastic read, recommend it!",
    "book_id": 1,
    "id": 1,
    "user_id": 1,
  },
  Object {
    "body": "Did not like it, expected much more...",
    "book_id": 2,
    "id": 2,
    "user_id": 1,
  },
]
`;

exports[`select query with unicode characters 1`] = `Array []`;

exports[`ts-implicit mode query 1`] = `
Array [
  Object {
    "author_id": 1,
    "categories": null,
    "id": 1,
    "name": "Black Swan",
    "rank": 1,
  },
]
`;
